replay/models/distributed_lightfm_wrap/lightfm_wrap.py

```diff
@@ -0,0 +1,733 @@
+            self._initialize_model()
+        self._initialize_world_size_and_threads()
+
+        if user_features is not None:
+            self.can_predict_cold_users = True
+        if item_features is not None:
+            self.can_predict_cold_items = True
+
+        log = log.select("user_idx", "item_idx", sf.lit(1).alias("relevance"))
+        interactions = log.repartition(self.world_size, "user_idx")
+
+        n_users, n_items = self._get_num_users_and_items(
+            log=interactions, previous_log=previous_log
+        )
+        self.max_seen_user_idx = n_users
+        self.max_seen_item_idx = n_items
+
+        user_features = self._convert_features_to_csr(log.select("user_idx").distinct(), user_features)
+        item_features = self._convert_features_to_csr(log.select("item_idx").distinct(), item_features)
+
+        (
+            user_features,
+            item_features,
+        ) = self.model._construct_feature_matrices(n_users, n_items, user_features, item_features)
+
+        for input_data in (
+            user_features.data,
+            item_features.data,
+        ):
+            self.model._check_input_finite(input_data)
+
+        if previous_log:
+            if self.model.item_embeddings:
+                self._reinitialize_embeddings(
+                    log,
+                    previous_log,
+                    num_users=n_users,
+                    num_items=n_items,
+                    user_features=user_features,
+                    item_features=item_features,
+                )
+            else:
+                raise ValueError(
+                    "Passed previous log will not affect training"
+                    "as the model has not been trained yet."
+                )
+
+        if self.model.item_embeddings is None:
+            self.model._initialize(
+                self.model.no_components,
+                item_features.shape[1],
+                user_features.shape[1],
+            )
+
+        if not item_features.shape[1] == self.model.item_embeddings.shape[0]:
+            raise ValueError("Incorrect number of features in item_features")
+        if not user_features.shape[1] == self.model.user_embeddings.shape[0]:
+            raise ValueError("Incorrect number of features in user_features")
+        if self.num_threads < 1:
+            raise ValueError("Number of threads must be 1 or larger.")
+
+        training_instance = LightFMTraining(self.model, self.world_size, self.num_threads)
+
+        host = training_instance.get_host_ip()
+        real_store = dist.TCPStore(
+            host,
+            self.tcp_port,
+            self.world_size + 1,
+            True,
+            timedelta(seconds=self.connection_timeout),
+        )
+
+        _tcp_port = self.tcp_port
+        _connection_timeout = self.connection_timeout
+        _num_epochs = self._num_epochs
+        _pygloo_timeout = self._pygloo_timeout_sec
+
+        def udf_to_map_on_interactions_with_index(
```

Comment:
Not a comment, but rather TODO for future: need to reconsider whole this function